legion_core/src/world.rs

```diff
@@ -903,6 +1178,122 @@ impl Default for World {
+        src_type: ComponentTypeId,
+        src_entities: &[Entity],
+        dst_entities: &[Entity],
+        src_data: *const u8,
+        dst_data: *mut u8,
+        num_components: usize,
+    );
+}
+
+/// Used along with CloneImpl, allows receiving results from a clone_from or clone_from_single call
+pub trait CloneImplResult {
+    /// For every entity that is copied, this function will be called, passing the entity in the
+    /// source and destination worlds
+    fn add_result(&mut self, src_entity: Entity, dst_entity: Entity);
+}
+
+/// Used along with CloneImpl, allows specifying that certain entities in the receiving world should
+/// be replaced with entities from the source world
+pub trait EntityReplacePolicy<'s> {
+    /// This function must return all entities in the source world that will replace data in the
+    /// destination world
+    fn src_entities<'a>(&'s self) -> Box<dyn Iterator<Item = Entity> + 'a>
+    where
+        's: 'a;
+
+    /// This function must return all entities in the destination world that will be replaced
+    fn dst_entities<'a>(&'s self) -> Box<dyn Iterator<Item = Entity> + 'a>
+    where
+        's: 'a;
+
+    /// When called, if the provided source entity is intended to replace an entity in the
+    /// destination world, the function must return the entity in the destination world, or None
+    fn get_dst_entity(&self, src_entity: Entity) -> Option<Entity>;
+}
+
+/// Used to opt-out of receiving results from a clone_from or clone_from_single call
+pub struct NoneCloneImplResult;
+impl CloneImplResult for NoneCloneImplResult {
+    fn add_result(&mut self, _src_entity: Entity, _dst_entity: Entity) {
+        // do nothing
+    }
+}
+
+/// Used to opt-out of replacing entities during a clone_from or clone_from_single call
+pub struct NoneEntityReplacePolicy;
+impl<'s> EntityReplacePolicy<'s> for NoneEntityReplacePolicy {
+    fn src_entities<'a>(&self) -> Box<dyn Iterator<Item = Entity> + 'a>
+    where
+        's: 'a,
+    {
+        Box::new(std::iter::Empty::default())
+    }
+
+    fn dst_entities<'a>(&self) -> Box<dyn Iterator<Item = Entity> + 'a>
+    where
+        's: 'a,
+    {
+        Box::new(std::iter::Empty::default())
+    }
+
+    fn get_dst_entity(&self, _src_entity: Entity) -> Option<Entity> { None }
+}
+
+/// Default implementation of CloneImplResult that uses a hash map
+pub struct HashMapCloneImplResult<'m>(pub &'m mut HashMap<Entity, Entity>);
+
+impl<'m> CloneImplResult for HashMapCloneImplResult<'m> {
+    fn add_result(&mut self, src_entity: Entity, dst_entity: Entity) {
+        self.0.insert(src_entity, dst_entity);
+    }
+}
+
+/// Default implementation of EntityReplacePolicy that uses a hash map
+pub struct HashMapEntityReplacePolicy<'m>(pub &'m HashMap<Entity, Entity>);
+
+impl<'m, 's> EntityReplacePolicy<'s> for HashMapEntityReplacePolicy<'m> {
+    fn src_entities<'a>(&'s self) -> Box<dyn Iterator<Item = Entity> + 'a>
```

Comment:
Is a box really needed? Can't we do something like 

```
 fn src_entities<'a>(&'s self) -> impl Iterator<Item = Entity> { 
....
```

Reply:
The box is required because impl Iterator is unsized